Align theme update JSON with theme pages on AMO

gijsk opened the pull request:
For the same theme data, the theme install page on AMO provides `"#"` as accent/textcolor strings, if those properties are not present. The fact that the update service then provides `null` is breaking themes in Firefox 60 and later. It seems sensible to align the two views of what is essentially the same data.

@eviljeff , can you confirm if there are current tests covering this part of the server? I didn't see any, but maybe I didn't look in the right place...

gijsk commented:
For context, this is related to https://bugzilla.mozilla.org/show_bug.cgi?id=1450975 . I don't see a github field to stick that link in, so...

EnTeQuAk commented:
Actually, there aren't any tests covering this specific use-case. The tests are in `src/olympia/addons/tests/test_theme_update.py` - let us know if you have the time to look at this too.

How do Firefox versions before 60 behave with this change? Are we certain they don't break somehow?

EnTeQuAk commented:
> After many years working on Mozilla stuff, "certain" seems like a strong word for me to use, but given that they seem to be fine with it if it's in the installation JSON on AMO to begin with, I wouldn't expect any problems... Does that help?

Still needs to be tested thoroughly at least in Firefox 59, Firefox 52 ESR and we'll have to make very sure this won't get backed out, changed or fixed in a later Firefox release since we'll have to support whatever Firefox 60 does for at least 3 years to come.

 Also, changing a value from `null` (doesn't exist) to `""` (empty value) in this specific case feels technically wrong.

gijsk commented:
Because of the braces it's changing to `"#"`, not `""`. I can make that more explicit if you prefer... :)

EnTeQuAk commented:
> Because of the braces it's changing to "#", not "". I can make that more explicit if you prefer... :)

That feels even more wrong, I can be soothed into accepting an empty string but `"#"` seems plain wrong to me.

gijsk commented:
Hm. So what if the database literally records `""` for a theme's accentcolor ? That would have the same output today (it gets substituted into `"#%s"` and thus produces `"#"`). In any case, this is also what the AMO theme page itself does. If you look at https://addons.mozilla.org/en-US/firefox/addon/flag-of-co-385587-2/ , the JSON string in there looks like this:

```
accentcolor: "#"
​author: "MozThemes"
​category: "Sports"
​description: null
​detailURL: "https://addons.mozilla.org/en-US/firefox/addon/flag-of-co-385587-2/"
​footer: "https://addons.cdn.mozilla.net/user-media/addons/385587/footer_persona_columbia.jpg?modified=31f490b6"
​footerURL: "https://addons.cdn.mozilla.net/user-media/addons/385587/footer_persona_columbia.jpg?modified=31f490b6"
​header: "https://addons.cdn.mozilla.net/user-media/addons/385587/Persona_columbia.jpg?modified=31f490b6"
headerURL: "https://addons.cdn.mozilla.net/user-media/addons/385587/Persona_columbia.jpg?modified=31f490b6"
iconURL: "https://addons.cdn.mozilla.net/user-media/addons/385587/preview_small.jpg?modified=31f490b6"
id: "385587"
name: "Flag of Colombia"
previewURL: "https://addons.cdn.mozilla.net/user-media/addons/385587/preview_large.jpg?modified=31f490b6"
textcolor: "#"
updateURL: "https://versioncheck.addons.mozilla.org/en-US/themes/update-check/385587"
version: "1.0"
```

Note the `textcolor` and `accentcolor` properties. I don't know what code produces that, but it's clearly already doing the same thing I'm proposing we do here.

Changing those pages to be consistent (and list `null` for those properties) would also be, err, consistent, but would break Firefox 60+ even more thoroughly (in that installing/previewing themes would break). Now it works until we do an add-on update...

eviljeff commented:
the install page JSON that gets passed to Firefox outputs "#" for a missing accentcolor/textcolor so, as crappy as that is, it _would_ be consistent to do the same in the update check.

ref https://addons.mozilla.org/api/v3/addons/addon/flag-of-co-385587-2/ (see `theme_data` - legacy frontend dumps theme_data as the install json too, afaik)

eviljeff commented:
more reference - the `"#"` comes from https://github.com/mozilla/addons-server/blob/master/src/olympia/addons/models.py#L1674:L1701 that just sticks a `#` on the front of the value.

eviljeff commented:
more reference - the `"#"` comes from https://github.com/mozilla/addons-server/blob/master/src/olympia/addons/models.py#L1674:L1701 that just sticks a `#` on the front of the value.

EnTeQuAk commented:
So, I do agree that the way the current addons API - returning `"#"` for an empty `accentcolor` (and others) value is wrong and should be changed to returning `null` as the theme-update service does.

That is something we can change, slowly and carefully but the API is primarily used by the addons-frontend and we have total control over that.

Changing the current - in my eyes more correct - `null` value for a not set `accentcolor` (and others) feels wrong and like a hack to me. Technically an empty string is possible and unfortunately for historical reasons we have `NULL` and an empty string in the database for `accentcolor` and all the other colour fields. So the only thing we can change is how we expose it.

Personally, I'm -1 changing our theme-update api because of a Firefox regression - especially for Firefox 60 since that locks us in to change it back to a technically more correct value for at least 3 years. I'd rather see the regression fixed in Firefox if possible.

cc @muffinresearch @krupa and @jvillalobos for additional opinions on this.

muffinresearch commented:
> Personally, I'm -1 changing our theme-update api because of a Firefox regression - especially for Firefox 60 since that locks us in to change it back to a technically more correct value for at least 3 years. I'd rather see the regression fixed in Firefox if possible.

Agreed, I think it's less risky to target the single area that has introduced the issue (Firefox) than to widen the potential impact by making a change on AMO.

eviljeff commented:
I think we shouldn't be serving different values for `textcolor`|`accentcolor` in the install data than the update service, so we should fix the API to return null in the `theme_data` block.

eviljeff commented:
Looks like the consensus (from the AMO side at least) is to not do this, in preference to fixing the regression in Firefox and changing the install JSON* to be correctly `null` for missing `textcolor` or `accentcolor` values.  * I'll file an issue.
